services/sn_analysis_service/app.py

```diff
@@ -112,55 +138,81 @@ async def run_sn_model(dd):
+            f = tempfile.NamedTemporaryFile(
+                suffix=".png", prefix="snplot_", dir="/tmp/", delete=False
+            )
+            f.close()
+            _ = sncosmo.plot_lc(
+                data,
+                model=fitted_model,
+                errors=result.errors,
+                model_label=source,
+                zp=23.9,
+                figtext=data_dict["resource_id"],
+                fname=f.name,
+            )
+            plot_data = base64.b64encode(open(f.name, "rb").read())
+            local_temp_files.append(f.name)
 
             # make some draws from the posterior (simulating what we'd expect
             # with an MCMC analysis)
             post = rng.multivariate_normal(result.parameters, result.covariance, 10000)
             post = post[np.newaxis, :]
             # create an inference dataset
-            dataset = az.convert_to_inference_data(
+            inference = az.convert_to_inference_data(
                 {x: post[:, :, i] for i, x in enumerate(result.param_names)}
             )
-            results.update({"data": {"posterior": dataset, "plot": plot_data}})
+            f = tempfile.NamedTemporaryFile(
+                suffix=".nc", prefix="inferencedata_", dir="/tmp/"
+            )
+            f.close()
+            inference.to_netcdf(f.name)
+            inference_data = base64.b64encode(open(f.name, 'rb').read())
+            local_temp_files.append(f.name)
+
+            result.update({"source": source, "fix_z": fix_z})
+
+            f = tempfile.NamedTemporaryFile(
+                suffix=".joblib", prefix="results_", dir="/tmp/"
+            )
+            f.close()
+            joblib.dump(result, f.name, compress=3)
+            result_data = base64.b64encode(open(f.name, "rb").read())
+            local_temp_files.append(f.name)
+
+            analysis_results = {
+                "inference_data": {"format": "netcdf4", "data": inference_data},
+                "plots": [{"format": "png", "data": plot_data}],
+                "results": {"format": "joblib", "data": result_data},
+            }
+
+            rez.update({"analysis": analysis_results})
+            rez.update(
+                {
+                    "status": "success",
+                    "message": f"Good results with chi^2/dof={result.chisq/result.ndof}",
+                }
+            )
         else:
-            results.update({"status": "failure", "message": "model failed to converge"})
+            log("Fit failed.")
+            rez.update({"status": "failure", "message": "model failed to converge"})
 
     except Exception as e:
-        results.update(
-            {"status": "failure", "message": f"problem running the model {e}"}
-        )
+        log(f"Exception while running the model: {e}")
+        log(f"{traceback.format_exc()}")
+        log(f"Data: {data}")
+        rez.update({"status": "failure", "message": f"problem running the model {e}"})
+    finally:
```

Comment:
Why prefer this over context managers?

Reply:
(For cleaning up files.)

Reply:
The `with TemporaryFile() as f` proved more finacky then I wanted, since I have to close the file before writing to it.